src/notebooksyncagent.cpp

```diff
@@ -782,82 +824,193 @@ bool NotebookSyncAgent::applyRemoteChanges()
-        KCalCore::Incidence::Ptr newIncidence = resources.at(i).incidence;
-        if (newIncidence.isNull()) {
-            continue;
-        }
-        if (mLocalDeletedUids.contains(newIncidence->uid())) {
-            LOG_DEBUG("Ignore incidence already deleted locally:" << resource.href);
-            continue;
+    if (incidence.isNull()) {
+        return false;
+    }
+    if (mLocalDeletedUids.contains(incidence->uid())) {
+        LOG_DEBUG("Ignore incidence already deleted locally:" << resource.href);
+        return false;
+    }
+
+    // find any existing incidence with this uid
+    QString incidenceIdentifier = QStringLiteral("%1 (%2)")
+            .arg(incidence->uid())
+            .arg(incidence->hasRecurrenceId() ? incidence->recurrenceId().toString() : QStringLiteral("original"));
+    mStorage->load(incidence->uid());
+    KCalCore::Incidence::Ptr storedIncidence;
+    switch (incidence->type()) {
+    case KCalCore::IncidenceBase::TypeEvent:
+        storedIncidence = mCalendar->event(incidence->uid(), incidence->hasRecurrenceId() ? incidence->recurrenceId() : KDateTime());
+        break;
+    case KCalCore::IncidenceBase::TypeTodo:
+        storedIncidence = mCalendar->todo(incidence->uid());
+        break;
+    case KCalCore::IncidenceBase::TypeJournal:
+        storedIncidence = mCalendar->journal(incidence->uid());
+        break;
+    case KCalCore::IncidenceBase::TypeFreeBusy:
+    case KCalCore::IncidenceBase::TypeUnknown:
+        qWarning() << "Unsupported incidence type:" << incidence->type();
+        return false;
+    }
+    if (storedIncidence) {
+        LOG_DEBUG("Updating existing event:" << incidenceIdentifier
+                                             << resource.href
+                                             << resource.etag);
+        storedIncidence->startUpdates();
+        IncidenceHandler::prepareImportedIncidence(incidence);
+        IncidenceHandler::copyIncidenceProperties(storedIncidence, incidence);
+        storedIncidence->setCustomProperty("buteo", "uri", resource.href);
+        storedIncidence->endUpdates();
+
+        // Save the modified incidence so it can be used to check whether there were
+        // local changes on the next sync.
+        mModifiedIncidenceICalData.insert(incidence->uid(), resource.iCalData);
+
+        // Save the incidence etag.
+        mUpdatedETags[incidence->uid()] = resource.etag;
+    } else {
+        LOG_DEBUG("Saving new event:" << incidenceIdentifier
+                                      << resource.href
+                                      << resource.etag);
+        KCalCore::Incidence::Ptr occurrence;
+        if (incidence->hasRecurrenceId()) {
+            // check to see if we had previously dissociated this occurrence from the recurring series.
+            occurrence = mCalendar->event(incidence->uid(), incidence->recurrenceId());
+            if (occurrence.isNull()) {
```

Comment:
Occurrence is fetched with similar line as storedIncidence above, so shouldn't it be guaranteed the check evaluates true?